### ISSUE TITLE
paddleocr output sequence length 40

### ISSUE THREAD

**Issue by HuyNgQc** [PaddlePaddle/PaddleOCR#17876]
### 🔎 Search before asking

- [x] I have searched the PaddleOCR [Docs](https://paddlepaddle.github.io/PaddleOCR/) and found no similar bug report.
- [x] I have searched the PaddleOCR [Issues](https://github.com/PaddlePaddle/PaddleOCR/issues) and found no similar bug report.
- [x] I have searched the PaddleOCR [Discussions](https://github.com/PaddlePaddle/PaddleOCR/discussions) and found no similar bug report.

### 🐛 Bug (问题描述)

Global:
  model_name: PP-OCRv5_server_rec # To use static model for inference.
  debug: true
  use_gpu: true
  epoch_num: 100
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output/PP-OCRv5_server_rec
  save_epoch_step: 1000
  eval_batch_step: [0, 123]
  cal_metric_during_train: true
  calc_epoch_interval: 1
  pretrained_model: /root/zorc/PaddleOCR/weightrs/PP-OCRv5_server_rec_pretrained.pdparams
  checkpoints:
  save_inference_dir:
  use_visualdl: false
  infer_img: doc/imgs_words/ch/word_1.jpg
  character_dict_path: /root/zorc/PaddleOCR/ppocr/utils/dict/ppocrv5_dict_mrz.txt
  max_text_length: &max_text_length 50
  infer_mode: false
  use_space_char: false
  distributed: true
  save_res_path: ./output/rec/predicts_ppocrv5.txt
  d2s_train_image_shape: [3, 48, 960]

Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Cosine
    learning_rate: 0.0005
    warmup_epoch: 2
  regularizer:
    name: L2
    factor: 3.0e-05

Architecture:
  model_type: rec
  algorithm: SVTR_HGNet
  Transform:
  Backbone:
    name: PPHGNetV2_B4
    text_rec: True
  Head:
    name: MultiHead
    head_list:
      - CTCHead:
          Neck:
            name: svtr
            dims: 120
            depth: 2
            hidden_dims: 120
            kernel_size: [1, 3]
            use_guide: True
          Head:
            fc_decay: 0.00001
      - NRTRHead:
          nrtr_dim: 384
          max_text_length: *max_text_length

Loss:
  name: MultiLoss
  loss_config_list:
    - CTCLoss:
    - NRTRLoss:

PostProcess:  
  name: CTCLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: MultiScaleDataSet
    ds_width: false
    data_dir: /root/zorc/full_data
    ext_op_transform_idx: 1
    label_file_list:
    - /root/zorc/full_data/train_split.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - RecAug:
        tia: false          # Tắt TIA transform
        deterioration: false
    - MultiLabelEncode:
        gtc_encode: NRTRLabelEncode
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_gtc
        - length
        - valid_ratio
  sampler:
    name: MultiScaleSampler
    scales: [[960, 48]]
    first_bs: &bs 64
    fix_bs: false
    divided_factor: [8, 16] # w, h
    is_training: True
  loader:
    shuffle: true
    batch_size_per_card: *bs
    drop_last: true
    num_workers: 16
Eval:
  dataset:
    name: SimpleDataSet
    data_dir: /root/zorc/full_data
    label_file_list:
    - /root/zorc/full_data/val_split.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - MultiLabelEncode:
        gtc_encode: NRTRLabelEncode
    - RecResizeImg:
        image_shape: [3, 48, 960]
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_gtc
        - length
        - valid_ratio
  loader:
    shuffle: false
    drop_last: false
    batch_size_per_card: 32
    num_workers: 4

I'm training the paddleocrv5 rec model and I'm encountering a problem where CTCLoss = 0 and I can't increment further. When checking in rec_ctc_loss, N=40 < the number of input characters. I've changed the scale value, increasing width resize from 960 to 1440, but T doesn't change.

[2026/03/30 07:44:13] ppocr INFO: train dataloader has 96 iters
[2026/03/30 07:44:13] ppocr INFO: valid dataloader has 49 iters
[2026/03/30 07:44:14] ppocr WARNING: The shape of model params head.ctc_head.fc.weight paddle.Size([120, 38]) not matched with loaded params head.ctc_head.fc.weight paddle.Size([120, 18385]) !
[2026/03/30 07:44:14] ppocr WARNING: The shape of model params head.ctc_head.fc.bias paddle.Size([38]) not matched with loaded params head.ctc_head.fc.bias paddle.Size([18385]) !
[2026/03/30 07:44:14] ppocr WARNING: The shape of model params head.gtc_head.embedding.embedding.weight paddle.Size([42, 384]) not matched with loaded params head.gtc_head.embedding.embedding.weight paddle.Size([18389, 384]) !
[2026/03/30 07:44:14] ppocr WARNING: The shape of model params head.gtc_head.tgt_word_prj.weight paddle.Size([384, 42]) not matched with loaded params head.gtc_head.tgt_word_prj.weight paddle.Size([384, 18389]) !
[2026/03/30 07:44:14] ppocr INFO: load pretrain successful from /root/zorc/PaddleOCR/weightrs/PP-OCRv5_server_rec_pretrained
[2026/03/30 07:44:14] ppocr INFO: During the training process, after the 0th iteration, an evaluation is run every 123 iterations
[2026/03/30 07:44:27] ppocr INFO: epoch: [1/100], global_step: 10, lr: 0.000012, acc: 0.000000, norm_edit_dis: 0.066948, CTCLoss: 0.000000, NRTRLoss: 4.004096, loss: 4.343603, avg_reader_cost: 0.24412 s, avg_batch_cost: 1.36741 s, avg_samples: 64.0, ips: 46.80367 samples/s, eta: 3:38:33, max_mem_reserved: 23355 MB, max_mem_allocated: 21876 MB
[2026/03/30 07:44:36] ppocr INFO: epoch: [1/100], global_step: 20, lr: 0.000025, acc: 0.000000, norm_edit_dis: 0.068300, CTCLoss: 0.000000, NRTRLoss: 3.391685, loss: 4.343603, avg_reader_cost: 0.00602 s, avg_batch_cost: 0.83359 s, avg_samples: 64.0, ips: 76.77658 samples/s, eta: 2:55:42, max_mem_reserved: 23355 MB, max_mem_allocated: 21876 MB

### 🏃‍♂️ Environment (运行环境)

OS ubuntu 22.04
Evironment VScode 
Python 3.10

### 🌰 Minimal Reproducible Example (最小可复现问题的Demo)

python /root/zorc/PaddleOCR/tools/train.py -c /root/zorc/PaddleOCR/configs/rec/PP-OCRv5/PP-OCRv5_server_rec.yml

**Comment by uniquefan** [comment]
The sequence length T is hardcoded to 40 during training, regardless of image width.

In ppocr/modeling/backbones/rec_pphgnetv2.py:1412-1416:

if self.text_rec:
    if self.training:
        x = F.adaptive_avg_pool2d(x, [1, 40])
    else:
        x = F.avg_pool2d(x, [3, 2])

In training mode, PPHGNetV2 with text_rec=True applies adaptive_avg_pool2d(x, [1, 40]), which always forces the output sequence to exactly 40 timesteps no matter how wide the input image is. That is why changing the scale width from 960 to 1440 had no effect on N.

Since your labels are up to 50 chars (max_text_length: 50), you get input_length (40) < label_length, so nn.CTCLoss (in ppocr/losses/rec_ctc_loss.py) returns 0 for these samples → no CTC gradient → CTCLoss = 0.000000 and acc = 0. Note the eval branch uses avg_pool2d(x, [3, 2]), which does scale with width – so training and eval behave differently.

How to fix it:

1. Make the training pooling width-dependent (match eval behavior):
   if self.training:
       x = F.avg_pool2d(x, [x.shape[2], 2])

   Then T = W / 16 (for 48px height input).

2. Make sure your image width gives T comfortably larger than your longest label. For ~50 chars use width ≥ 640 (T = 80), e.g. scales: [[640, 48]] and d2s_train_image_shape: [3, 48, 640].

3. Keep max_text_length consistent with that width (e.g. 128 is fine, but 50 works too as long as T > 50).

With the fixed-width training pool, samples whose label length is close to T will also have very little room for CTC blanks, so leaving a margin (T ≈ 2× max label length) gives much better results.
